### DISCUSSION TITLE
Cmake presets and cmake.sourceDirectory array

### DISCUSSION THREAD

**Discussion by jankap** [microsoft/vscode-cmake-tools#3059] [Q&A] [answered]
Hi :)

In #1374 the `${sourceDir}` variable was added which can be an array. I was wondering how that plays together with cmake presets, using `CMakePresets.json`? 

What happens if `binaryDir` is set to `"${sourceDir}/out"` and `sourceDir` is an array?

What I'd like to achieve is to support two different build systems, both relying on CMake. One for testing, the other for building the code.

root
-- CmakeLists.txt <- test root 
---- test/CmakeLists.txt <- test definition
---- build-src/CmakeLists.txt <- build src

The "test root" CmakeLists.txt does not include the one in build-src. I'd like to handle these completely independent and add two cmake config and build presets.

Is that possible?

**Comment by bobbrow** [comment, marked as answer]
When `${sourceDir}` is used in CMakePresets.json, it refers to the path to the project's source directory (e.g. CMAKE_SOURCE_DIR). Reference: https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html#macro-expansion

We just happen to support a similarly named variable for consistency when resolving settings (most notably for `cmake.buildDirectory`) because using `${workspaceRoot}` could result in path collisions when generating multiple projects in the same folder. You can use `${sourceDir}` and `${sourceDirectory}` interchangeably.

For your project layout, it seems to me that you could set `cmake.sourceDirectory` to the two root CMakeLists.txt that you have and the extension should allow you to manage both of them independently (with separate presets files and everything).  Something like this:
```json
  "cmake.sourceDirectory": [
    "${workspaceFolder}",
    "${workspaceFolder}/build-src"
  ]
```
